lib/connection.ts

```diff
@@ -577,7 +577,9 @@ export class Connection extends Entity {
 
       onAbort = () => {
         removeListeners();
-        rheaSession.close();
+        if (rheaSession.is_open()) {
+          rheaSession.close();
```

Comment:
Why wouldnt rhea consider this to be a no-op if the session is not open in the first place?
Is it because `is_open` checks for both local and remote open while `close()` ends up in https://github.com/amqp/rhea/blob/master/lib/endpoint.js#L48 where we check only for local open?

Reply:
Right. Because close doesn't check if the remote is open, we end up trying to close something that hasn't actually been opened on the remote side yet.

Reply:
How do we ensure "local" stuff get cleaned up in such cases?

Reply:
@grs What is your recommendation here? Should we always check `is_open()` before attempting a `close()` or should `close()` be smart enough to not communicate to the server to close something that is not open?

Reply:
> How do we ensure "local" stuff get cleaned up in such cases?

I could call `remove()` on the rhea receiver/sender, which should remove it from the session. There is a note that this shouldn't be done on active links, but I don't think the aborted links should be considered active.

I can do some more testing to see what happens if I abort these without closing the connection (in our libraries we do close the connection so nothing is left open).

Reply:
The close() does involve a check on whether the endpoint is locally open or not, so I don't believe a check is necessary to prevent a close for a session that has already been opened and closed. Closing locally is fine even if remote has not yet signalled open. What would not be ok is sending a close when no open has been sent (i.e. calling close() if open()/begin() was never called). The only way I see that could happen is if the session object was never opened locally.

The session returned by rhea from Connection.create_session() is not locally opened. However the links returned from open_sender/open_receiver are opened.

Reply:
I think the local workaround is fine for now, although it'd be nice to put it into a function and put a comment on it to indicate why it's needed (I can easily see someone thinking - oh, we don't need to bother and never be able to trigger the edge case!).

Also, can we add a test for this?

Reply:
So an alternative I'm thinking about for receiver and sender (might work for session as well) is instead of calling `rheaReceiver.close()` right away in our `onAbort` handler, I can attach a "receiver_open" event listener on the `rheaReceiver` object. If it is invoked, this would mean the receiver was opened after we initiated our abort, and at this point we just close the receiver.

We could also have the is_open check I've added and close if it's true, since at that point we know the remote is open as well.